beacon_node/beacon_processor/src/lib.rs

```diff
@@ -762,205 +456,139 @@ impl<T: BeaconChainTypes> BeaconProcessorSend<T> {
                     &metrics::BEACON_PROCESSOR_SEND_ERROR_PER_WORK_TYPE,
                     &[work_type],
                 );
-                Err(Box::new(e))
+                Err(e)
             }
         }
     }
 }
 
-/// A consensus message (or multiple) from the network that requires processing.
-#[derive(Derivative)]
-#[derivative(Debug(bound = "T: BeaconChainTypes"))]
-pub enum Work<T: BeaconChainTypes> {
+pub type AsyncFn = Pin<Box<dyn Future<Output = ()> + Send + Sync>>;
+pub type BlockingFn = Box<dyn FnOnce() + Send + Sync>;
+pub type BlockingFnWithManualSendOnIdle = Box<dyn FnOnce(SendOnDrop) + Send + Sync>;
+
+/// Indicates the type of work to be performed and therefore its priority and
+/// queuing specifics.
+pub enum Work<E: EthSpec> {
     GossipAttestation {
-        message_id: MessageId,
-        peer_id: PeerId,
-        attestation: Box<Attestation<T::EthSpec>>,
-        subnet_id: SubnetId,
-        should_import: bool,
-        seen_timestamp: Duration,
+        attestation: GossipAttestationPackage<E>,
+        process_individual: Box<dyn FnOnce(GossipAttestationPackage<E>) + Send + Sync>,
+        process_batch: Box<dyn FnOnce(Vec<GossipAttestationPackage<E>>) + Send + Sync>,
     },
     UnknownBlockAttestation {
-        message_id: MessageId,
-        peer_id: PeerId,
-        attestation: Box<Attestation<T::EthSpec>>,
-        subnet_id: SubnetId,
-        should_import: bool,
-        seen_timestamp: Duration,
+        process_fn: BlockingFn,
     },
     GossipAttestationBatch {
-        packages: Vec<GossipAttestationPackage<T::EthSpec>>,
+        attestations: Vec<GossipAttestationPackage<E>>,
+        process_batch: Box<dyn FnOnce(Vec<GossipAttestationPackage<E>>) + Send + Sync>,
     },
     GossipAggregate {
-        message_id: MessageId,
-        peer_id: PeerId,
-        aggregate: Box<SignedAggregateAndProof<T::EthSpec>>,
-        seen_timestamp: Duration,
+        aggregate: GossipAggregatePackage<E>,
+        process_individual: Box<dyn FnOnce(GossipAggregatePackage<E>) + Send + Sync>,
+        process_batch: Box<dyn FnOnce(Vec<GossipAggregatePackage<E>>) + Send + Sync>,
     },
     UnknownBlockAggregate {
-        message_id: MessageId,
-        peer_id: PeerId,
-        aggregate: Box<SignedAggregateAndProof<T::EthSpec>>,
-        seen_timestamp: Duration,
+        process_fn: BlockingFn,
     },
     UnknownLightClientOptimisticUpdate {
-        message_id: MessageId,
-        peer_id: PeerId,
-        light_client_optimistic_update: Box<LightClientOptimisticUpdate<T::EthSpec>>,
-        seen_timestamp: Duration,
+        parent_root: Hash256,
```

Comment:
Some of the `WorkItems` have some extra metadata (`parent_root`, in this case) to help with logging and reprocessing operations.